pkg/faketikv/raft.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package faketikv
+
+import (
+	"math/rand"
+	"sort"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/pd/pkg/faketikv/cases"
+	"github.com/pingcap/pd/pkg/faketikv/simutil"
+	"github.com/pingcap/pd/server/core"
+)
+
+// RaftInfo records all raft infomation.
+type RaftInfo struct {
+	*core.RegionsInfo
+	conn         *Conn
+	regionchange map[uint64][]uint64
+}
+
+// NewRaftInfo creates the initialized raft with the configuration.
+func NewRaftInfo(conf *cases.Conf, conn *Conn) (*RaftInfo, error) {
+	r := &RaftInfo{
+		RegionsInfo:  core.NewRegionsInfo(),
+		conn:         conn,
+		regionchange: make(map[uint64][]uint64),
+	}
+
+	splitKeys := generateKeys(len(conf.Regions) - 1)
+	for i, region := range conf.Regions {
+		meta := &metapb.Region{
+			Id:          region.ID,
+			Peers:       region.Peers,
+			RegionEpoch: &metapb.RegionEpoch{ConfVer: 1, Version: 1},
+		}
+		if i > 0 {
+			meta.StartKey = []byte(splitKeys[i-1])
+		}
+		if i < len(conf.Regions)-1 {
+			meta.EndKey = []byte(splitKeys[i])
+		}
+		regionInfo := core.NewRegionInfo(meta, region.Leader)
+		regionInfo.ApproximateSize = region.Size
+		regionInfo.ApproximateKeys = region.Keys
+		r.SetRegion(regionInfo)
+	}
+
+	return r, nil
+}
+
+func (r *RaftInfo) stepRegions(c *ClusterInfo) {
```

Comment:
maybe let the config as an inner config will be better.

Reply:
+1

Reply:
There is an another function named `allocID` in `stepSplit` so that we need to use `ClusterInfo`.